Add option to keep manually selected samplers on API Type switch

leandrojofre opened the pull request:
## Changes - Issue #4671
Why? Presets in Text Completion are often bound to an API. In example: the user may use X preset with Y API through Connections Profiles, only using in reality just a few samplers from the dozens the preset has. Issue: if the user changes the Connection Profile or API Type, the displayed samplers are reset to show everything supported by the API, eliminating any manual display of actual samplers used by the user. Even if both API Types are using the same TC preset, the samplers are reset anyway. This PR tries to add an option for advanced users to have in hand the samplers they want by making the sampler selector popup more permanent.
- Added the capability to save manually selected samplers by the user from the Sampler Select popup.
- It is stored individually per TC preset, on a localforage instance.
- It can be turned `on`/`off` from the popup.
- When the feature is `off`, it uses ST's default behavior for showing/hiding sampler parameters.
- A warning tooltip was added warning the users that while the feature is active, ST won't show API specific parameters automatically - See [Here](https://github.com/SillyTavern/SillyTavern/pull/4692/files#diff-78172ab44f6a372ef2db77c374addfda6907fc6478d3737b8c7df5cfbe792fc7R11).

## Commits
- [Feat](https://github.com/SillyTavern/SillyTavern/commit/3d73605ab49eca7666319a0167daa5bbac01eb35) - Save and load manually selected samplers per preset
- [Fix](https://github.com/SillyTavern/SillyTavern/commit/3f1d75c230fef5a1e7c0f71a91f2ad1d07335310) - Sync sampler popup with selected sampler state settings
- [Fix](https://github.com/SillyTavern/SillyTavern/commit/f59fd72ab600921b81f67b64e3976b80608ecf12) - Prevent breaking visualization after disabling keep selected samplers
- [Fix](https://github.com/SillyTavern/SillyTavern/commit/bbdc04b3a89cc3c52e5cf9f441bb4c532927d620) - ESLint errors in samplerSelect.js
- [Fix](https://github.com/SillyTavern/SillyTavern/commit/566d8921040a8f167c4d68759e526f34ca65b89c) - ESLint errors in textgen-settings.js
- [Docs](https://github.com/SillyTavern/SillyTavern/commit/f5fd03ef658d82e4162fd13de572cc15db3c6f2d) - Write JSDocs for new methods in textgen-settings.js
- [Update](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/16239ac237327f2fa59c70845eee1ac789c5f45d) - Make the reset button work individually per preset
> It didn't make sense to make it wipe the manually selected state for all presets. It was made at the start while the feature still wasn't intended to store the selected state per preset.
- [Fix](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/28a25b60e63abd5d29d308fde9388d3ad41d901d) - Trailing whitespace at the end of class selector
- [Fix](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/567a3d556d6847137934185f80248b9443de2302) - Update tooltip information for force samplers toggle
- [Fix](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/bfd7eb3ea925133ee826602562c5650f5fd0c61a) - Prevent calling isSamplerManualPriorityEnabled a bajillion times
- [Fix](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/0dd2fce07de65bdb655dcd2526bf4ad18d2fb796) - Unbrick switch from manually-forced to non-manually-forced preset
> This prevents samplers non controlled by `[data-tg-type]` from being sent to the shadow realm if they were previously hidden by `[data-tg-samplers]`. `is_preset_switch` is just to prevent breaking muscle memmory of users that will not use force manually selected samplers.
- [Fix](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/cd59af6ad4424dc98d35c2795875c7bc8c8144d3) - Load manually selected samplers on connection profile switch
> Changing preset updates samplers ONLY IF force manually selection is enabled. Changing API before preset would update the samplers, but using the settings of the previous preset, due to preset being changed AFTER the API.
- [Fix](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/2da37d5c82fe91a6b01f8a8eca7e44208200cc8f) - Switch API Type controls in the API Connections panel
- [Fix](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/710d0e06bf81a81f3bbb34d29cf27a26732f91d6) - ESLint errors in textgen-settings.js
- [Fix](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/6fbd5d0e94a0338cab6a378f1df37e763da9184e) - Update sampler selection only with textgenerationwebui as main API selected

## Checklist:

- [X] I have read the [Contribution guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

leandrojofre committed:
Feat - Save and load manually selected samplers per preset

leandrojofre committed:
Fix - Sync sampler popup with selected sampler state settings

leandrojofre committed:
Fix - Prevent breaking visualization after disabling keep selected samplers

leandrojofre committed:
Fix - ESLint errors in samplerSelect.js

leandrojofre committed:
Fix - ESLint errors in textgen-settings.js

leandrojofre committed:
Docs - Write JSDocs for new methods in textgen-settings.js

leandrojofre committed:
Update - Make the reset button work individually per preset

leandrojofre commented:
/gemini summary

Cohee1207 commented:
@codex need a test plan, no idea how to properly approach it.

leandrojofre committed:
Fix - Trailing whitespace at the end of class selector

Cohee1207 commented:
Man, this codex dude never followed up. And I still don't know how to test this. Can you pinky promise that this works and has no bugs?

leandrojofre commented:
@Cohee1207 lmao, the bot just gave up.
I've been using this branch in my main install since the launching of this PR without finding any bugs, but I don't feel confident with a pinky promise. In theory, the preset switch and the API Type switch should work the same as they do now in staging for users who have not interacted with the [toggle button](https://github.com/SillyTavern/SillyTavern/pull/4692/files#diff-78172ab44f6a372ef2db77c374addfda6907fc6478d3737b8c7df5cfbe792fc7R8-R10) of this feature from the samplers popup.
I can try use it a few days more and confirm it has no bugs by this weekend. Tho, I don't have issues waiting if you want someone else to test first.

leandrojofre commented:
Correct me if I'm wrong: Does hiding samplers actually deletes the sampler from the request body, or the request body always has all the samplers the API requires and the left menu is just visual? Just to know if I should modify [this message](https://github.com/SillyTavern/SillyTavern/pull/4692/files#diff-78172ab44f6a372ef2db77c374addfda6907fc6478d3737b8c7df5cfbe792fc7R11). I could swear that's not the case, but a message from Ross in the code channel on Discord confused me 🤣

Cohee1207 commented:
> Does hiding samplers actually deletes the sampler from the request body

No, sampler selection was always a purely and entirely visual thing. That's why I never understood its significance.

leandrojofre commented:
Oh nice, just visual. I'll modify the warning then, ty.

> That's why I never understood its significance.

I find the menu very nice when you use only the same two or three samplers regardless of the API and ignore the rest; it helps reduce bloat in the left menu and visual pain finding X thing. But it always bugged me that the thing resets on API change.

leandrojofre committed:
Fix - Update tooltip information for force samplers toggle

leandrojofre committed:
Fix - Prevent calling isSamplerManualPriorityEnabled a bajillion times

leandrojofre committed:
Fix - Unbrick switch from manually-forced to non-manually-forced preset

This prevents samplers non controlled by [data-tg-type] from being sent to the shadow realm if they were previously hidden by [data-tg-samplers]. is_preset_switch is just to prevent breaking muscle memmory of users not using force manually selected samplers.

leandrojofre committed:
Fix - Load manually selected samplers on connection profile switch

Changing preset does not update samplers ONLY IF force manually selection is disabled. Changing API before preset would update the samplers, but using the settings of the previous preset, due to preset being changed AFTER the API.

RossAscends commented:
> Does hiding samplers actually deletes the sampler from the request body

As Cohee said, the sampler select is just visual to clear clutter from the sampler panel and hopefully enable you to view all your used samplers without needing to scroll the panel. I made a separate extension called Sampler Killswitches to actually remove the params from the outgoing API Call. 

I've not tested the branch yet, but looking through the code, one thing I'm not crazy about conceptually here is linking the displayed samplers to a preset. What preset is this? TC sampler presets?

I could understand it being linked to a particular API (i.e. likely via connection profiles as it seems you've done), since that is what defines which samplers can/should be displayed and used.

leandrojofre commented:
@RossAscends 
> What preset is this? TC sampler presets?

Yup, Text Completion sampler presets.

> I could understand it being linked to a particular API [...] since that is what defines which samplers can/should be displayed and used.

The idea of this is pretty much ignore which samplers are used by the API and only show the ones selected by the user. Trying to make an example outside connection profiles:
- Lets say a user only uses temperature, top_k and min_p; the rest of the samplers are always in their respective neutralized values and are never changed. The user took its time ticking off each and every of the samplers not used to clean the panel from unused things. If the user wants to change from Koboldcpp to OpenRouter because wathever they're doing requires a single call from a model their computer can't run, the samplers in the panel would get reset and show again all the list of samplers they were not using.

Why not API tied? I would point out two things mainly.
- First, it gives more freedom; you can have different layouts for a single API or share a single layout across multiple APIs.
- Second, there are APIs that that might need more or less samplers depending on the model you're connected; biggest example is OpenRouter, were you have models that are unresponsive to top_p/min_p, or that might or not use include_reasoning, banned_tokens might nuke the model, be needed or ignored, etc.

In the way this thing is made, if a preset has the toggle OFF to use manual selection it will fallback to display the samplers used by the API the next time the panel updates - Just as it works right now. To be clear, it is **not** trying to replace current behaviour, it adds an additional one. By default, the toggle to force manual selection is OFF, the user would have to turn it ON for it to make an effect (I should probably change the button icon to a lock).

Please point out if you think that it still needs to be API tied, or ask wathever you're in doubt while you test - I'm open to change it.

leandrojofre committed:
Fix - Switch API Type controls in the API Connections panel

leandrojofre committed:
Fix - ESLint errors in textgen-settings.js

leandrojofre committed:
Fix - Update sampler selection only with textgenerationwebui as main API selected

leandrojofre commented:
/gemini review

leandrojofre commented on public/scripts/textgen-settings.js:
```diff
@@ -1064,6 +1181,41 @@ export function initTextGenSettings() {
+function showSamplerControls(api_type = null, is_preset_switch = false) {
+    const prioritizeManualSamplerSelect = isSamplerManualPriorityEnabled();
+
+    if (is_preset_switch && !prioritizeManualSamplerSelect) return;
```
Preset switch is not meant to toggle samplers display if this new feature is OFF, same as it worked before this PR. Sampler display toggle when the feature is OFF is already handled with `$('#textgen_type').on('change')`.

leandrojofre commented:
> [...] and confirm it has no bugs by this weekend.

I've been using this branch in my install, switching between TC and CC with different APIs, and it seems to be working well. All the bugs I could find have been fixed.

leandrojofre committed:
Styles - Change the sampler selection lock button icon to an actual lock.

leandrojofre committed:
Merge branch 'staging' into 1-keep-selected-samplers-on-switch

Wolfsblvt commented:
@codex review this and make a detailed test plan on how to test this.

Wolfsblvt commented:
@codex review this and make a detailed test plan on how to test this.

Wolfsblvt commented:
Man, really have to do everything myself. I thought codex was magic, eh?

@Cohee1207 here is your test plan, as written by codex on a run against the repo with this PR.

------------------------

# PR 4692 Test Plan: Persisting Manual Sampler Selection per Preset

## Scope
Focus on the new ability to pin sampler controls per Text Completions preset, the accompanying manual-priority toggle in the sampler selector popup, the persistence layer backed by `localforage`, and regressions in sampler visibility, preset switching, and the Connection Manager extension. Also validate the adjusted OpenAI `max_tokens` limit in the UI.

## Pre-requisites
1. `git fetch origin staging` and `git checkout pr4692` to test against the PR branch.
2. Run `npm install` (only once if node_modules already present).
3. Launch SillyTavern with `npm run start` and open it in a Chromium-based browser with devtools access.
4. Ensure at least one Text Completions preset exists (the default "Default" preset is fine) and that the Text Completions API is reachable (the tests can use mock/offline mode because they mostly exercise UI logic).

## Manual Verification

### A. Baseline sampler behaviour (regression guard)
1. Open **Settings → Text Generation → Text Completions**.
2. Without opening the sampler popup, switch between several API types (e.g. `OpenAI`, `KoboldCpp`, `textgenerationwebui`).
3. Confirm that sampler blocks tagged for each API via `data-tg-type` / `data-tg-type-mode` still appear/disappear automatically when the API type changes.
4. Confirm there are no JS errors in the browser console.

### B. Persisting manual sampler selections per preset
1. Set **Main API** to `textgenerationwebui`.
2. Open the sampler selector popup (gear icon) and observe the new lock icon (`Prioritize...`). Verify the tooltip copy.
3. Toggle the lock **on**. Check that the icon gains the `toggleEnabled` class (visual change) and that no console errors appear.
4. Enable a subset of samplers (e.g. Temperature, Top P, Min P) and disable several others.
5. Close the popup. Confirm that only the manually enabled sampler panels remain visible in the left settings column.
6. Reload the page. Verify that, after settings load, the same subset of samplers is shown and the popup lock remains toggled on. (Confirms `localforage` persistence and the async loading path.)
7. Switch to a different preset (or create a new preset). Verify that the new preset starts with default sampler visibility and an independent lock state. Return to the original preset and confirm its manual selection is still respected.

### C. Resetting manual selections
1. With manual priority still enabled for the preset, reopen the sampler popup.
2. Click **Reset custom sampler selection**.
3. Confirm the sampler checkboxes return to the default state (matching the auto-selected controls for the current API type) and the lock icon returns to the off state.
4. Close the popup and verify the main settings column shows the full default sampler set again.
5. Refresh the page to ensure the cleared state persists.
6. Re-open the popup to confirm the lock remains off and the manual selections are gone (no `selectedSamplers` stored for the preset in `IndexedDB → localforage → SillyTavern_TextCompletions`).

### D. Behaviour when manual priority is off
1. With the lock off, enable or disable some samplers in the popup.
2. Switch API types in the left column (e.g. to `KoboldCpp` and back to `textgenerationwebui`).
3. Confirm that sampler visibility returns to the automatic defaults (manual tweaks are ignored unless the lock is on).
4. Ensure the popup still reflects the defaults and that `selectedSamplers` for the preset stays empty in localforage.

### E. Handling empty manual selections
1. Turn the lock **on**, but deselect every sampler in the popup (so nothing is explicitly enabled).
2. Close the popup.
3. Verify that the settings page falls back to the default sampler display (because no manual entries exist) and that no samplers disappear permanently.
4. Reload the page and confirm the behaviour persists (no hidden controls due to an empty manual list).

### F. Interaction with Connection Manager extension
1. Enable the Connection Manager extension, create a new profile that targets Text Completions, and record commands for `preset`, `api`, and at least one manual sampler selection (lock on).
2. Switch to another profile (or default state) to reset the UI.
3. Reapply the saved profile.
4. Confirm that the preset is set first, the API follows, and the manually pinned samplers become visible without needing to reopen the popup.
5. Inspect the browser console to ensure no errors fire during profile apply.

### G. Storage persistence validation
1. Open browser devtools → Application tab → IndexedDB → `localforage` → `SillyTavern_TextCompletions`.
2. Confirm that `selectedSamplers` stores one entry per preset with keys for each sampler plus the `st_manual_priority` flag.
3. Delete the entry manually and refresh SillyTavern. Verify that the UI rebuilds with defaults (matching the reset workflow).

### H. Regression around OpenAI max token slider
1. Switch Main API to `OpenAI`.
2. Inspect the "Max Response Length (tokens)" field and verify that the HTML `max` attribute is now `65536`.
3. Try to type a value larger than `65536` and confirm the browser prevents it.

## Optional Explorations
- Repeat the persistence workflow on additional presets to ensure cross-preset isolation works when multiple presets have manual priorities enabled simultaneously.
- Evaluate performance of `showSamplerControls` when many samplers are toggled to ensure there are no perceptible UI stalls.
- Temporarily block IndexedDB (Chrome devtools > Application > Clear storage with the tab open) and confirm the UI falls back to default sampler visibility with console warnings, not crashes.

## Expected Outcomes
- Manual sampler selections persist per preset only when the lock is active, and they survive reloads via localforage.
- Default sampler visibility logic continues to work for all API types when manual priority is off.
- Resetting or clearing storage restores defaults without leaving the UI in an inconsistent state.
- Connection Manager replays commands in the new order without errors and honours manual sampler visibility.
- The OpenAI max token field enforces the 65,536 upper bound.

leandrojofre committed:
Fix - Await for sampler selection reset to finish

leandrojofre committed:
Update - Normalize function name

leandrojofre committed:
Update - Preserve manual sampler selections on preset renaming

Wolfsblvt commented:
Let us know once this is stable, if you are currently still adding commits/changing stuff.

leandrojofre commented:
@Wolfsblvt I'm following codex test plan and fixing found bugs, there are some cases I didn't remember existed. I'll leave a comment once it's finished.

leandrojofre committed:
Fix - Properly reset sampler manual selections

leandrojofre committed:
Update - Ensure selection reset displays API specific samplers

leandrojofre committed:
Fix - ESLint errors

leandrojofre commented:
Finished the bug fixing and used the branch a little. It is ready if anyone else wants to test now.

leandrojofre committed:
Fix - Avoid modifying non-existent data

leandrojofre commented:
Last commit is tested, I forgot to upload it yesterday.

Cohee1207 commented:
@codex review

Cohee1207 commented on public/scripts/textgen-settings.js:
```diff
@@ -357,6 +362,123 @@ export function getTextGenServer(type = null) {
+/**
+ * Initializes the configuration object for manually selected samplers.
+ */
+async function loadPresetSelectedSamplers() {
```
I think it's safe to say that these functions belong to samplerSelect.js
It will be nice to move them there.

leandrojofre replied:
@Cohee1207 Okay. So, do I move these?
- `loadPresetSelectedSamplers`
- `savePresetSelectedSamplers`
- `resetPresetSelectedSamplers`
- `setPresetSamplersState`
- `getManualPresetSamplers`
- `getManualActivePresetSamplers`
- `toggleSamplerManualPriority`
- `isSamplerManualPriorityEnabled`

leandrojofre replied:
@Cohee1207 Done, check [d7cdb26](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/d7cdb26753ae22733bfb5cedbd0011f1a46e3c5c). Also check [7c590cf](https://github.com/SillyTavern/SillyTavern/pull/4692/commits/7c590cfe21ceb522761840ada856d6130e033ece), there's a small change in `samplerSelect.js` to prevent a circular depency error generated by moving the things to that file and importing from `textgen-settings.js`.

Wolfsblvt replied:
Sounds about right. The more we split functions into logical groups, the easier those files will be to read and maintain, tbh.

leandrojofre commented on public/scripts/samplerSelect.js:
```diff
@@ -307,6 +336,7 @@ async function listSamplers(main_api, arrayOnly = false) {
         };
         console.log(sampler, targetDOMelement.prop('id'), isInDefaultState(), isInForceShownArray, isInForceHiddenArray, shouldBeChecked());
         if (displayname === undefined) { displayname = sampler; }
+        if (main_api === 'textgenerationwebui') setPresetSamplersState(sampler, shouldBeChecked());
```
This is made to keep localforage data synchronized with user selections even when the feature is off.

leandrojofre commented on public/scripts/textgen-settings.js:
```diff
@@ -1069,6 +1195,41 @@ export function initTextGenSettings() {
+function showSamplerControls(api_type = null, is_preset_switch = false) {
+    const prioritizeManualSamplerSelect = isSamplerManualPriorityEnabled();
+
+    if (is_preset_switch && !prioritizeManualSamplerSelect) return;
```
When a preset switches (`selectPreset`), it shouldn't do anything by default, so if there's a switch but the feature is off, do nothing `is_preset_switch && !prioritizeManualSamplerSelect`. If the feature is `on` in the new preset activated, then we need to hide/show the samplers selected in that preset.

leandrojofre committed:
Update - Move manual sampler control methods to samplerSelect.js

leandrojofre committed:
Fix - Break circular file dependency between textgen-settings.js and samplerSelect.js

setting_names was unavailable during page load, throwing an error and preventing ST from loading.

Cohee1207 committed:
Merge branch 'staging' into 1-keep-selected-samplers-on-switch

Cohee1207 committed:
Update menu layout

Cohee1207 committed:
Remove pointless await

Cohee1207 committed:
Update param naming to camelCase

Cohee1207 reviewed (approved):
I clicked around and it seems to work. I did not fully understand the relation between the selected samplers and presets, but I hope you can provide a documentation update for this.
